File: ios/Yappy/Data/E2E.swift
import Foundation

/// The session layer: who a message gets sealed to, what this device can open,
/// and where the answer is kept.
///
/// The envelope is in `Cipher` and the ratchet under it in `Ratchet`. What lives
/// here is everything that has to touch the network or the disk: claiming the
/// prekeys that start a session, fetching the identity key a signature is
/// checked against, and writing down what a message said — because with a
/// ratchet, that is the only copy that survives.
actor E2E {
    private static let flagKey = "yappy.e2e.conversations"
    /// Long enough that reading a screenful is one request per person.
    private static let directoryTTL: TimeInterval = 5 * 60

    private let repo: YappyRepository
    private let session: SessionStore
    private let keys: DeviceKeys
    private let store: E2EStore

    private struct DirectoryEntry {
        let fetchedAt: Date
        /// device id → Ed25519 identity key.
        let byDevice: [String: String]
    }

    private var directory: [String: DirectoryEntry] = [:]

    init(repo: YappyRepository, session: SessionStore, keys: DeviceKeys, store: E2EStore) {
        self.repo = repo
        self.session = session
        self.keys = keys
        self.store = store
    }

    /// Two locks: the build, and the per-conversation flag.
    nonisolated var available: Bool {
        #if DEBUG
            return true
        #else
            return false
        #endif
    }

    nonisolated func isPrivate(_ conversationId: String) -> Bool {
        guard available else { return false }
        let flagged = UserDefaults.standard.stringArray(forKey: Self.flagKey) ?? []
        return flagged.contains(conversationId)
    }

    nonisolated func setPrivate(_ conversationId: String, _ on: Bool) {
        var flagged = Set(UserDefaults.standard.stringArray(forKey: Self.flagKey) ?? [])
        if on { flagged.insert(conversationId) } else { flagged.remove(conversationId) }
        UserDefaults.standard.set(Array(flagged), forKey: Self.flagKey)
    }

    // ── the devices of everybody else ────────────────────────────────────────

    /// Everything the directory says about one person's devices.
    ///
    /// Refetched when a message names a device the cache has never seen — that
    /// is exactly what somebody adding a phone looks like from here, and the
    /// alternative is that their first message from it is permanently
    /// unreadable.
    private func devicesOf(_ userId: String, wanted: String? = nil) async -> [String: String] {
        if let hit = directory[userId],
           Date().timeIntervalSince(hit.fetchedAt) < Self.directoryTTL,
           wanted == nil || hit.byDevice[wanted!] != nil
        {
            return hit.byDevice
        }
        do {
            let devices = try await repo.userKeys(userId).devices
            let byDevice = Dictionary(
                devices.map { ($0.deviceId, $0.identityKey) },
                uniquingKeysWith: { first, _ in first }
            )
            directory[userId] = DirectoryEntry(fetchedAt: Date(), byDevice: byDevice)
            return byDevice
        } catch {
            return directory[userId]?.byDevice ?? [:]
        }
    }

    /// The identity key a device publishes, which is what its signatures are checked against.
    private func identityKey(of userId: String, device deviceId: String) async -> String? {
        await devicesOf(userId, wanted: deviceId)[deviceId]
    }

    // ── sealing ──────────────────────────────────────────────────────────────

    /// What a private send puts on the wire: one ciphertext per recipient
    /// device, each under its own ratchet.
    ///
    /// Own devices included — a message sent from a phone has to be readable on
    /// the iPad — but not the device doing the sending, which cannot hold a
    /// ratchet session with itself and writes down what it said instead.
    ///
    /// Nil means there was nobody to encrypt to, which is not an error: the
    /// caller sends in the clear rather than posting something nobody can read.
    /// A device whose signed prekey does not verify is dropped on its own —
    /// that is one bad device, and everybody else is still owed their copy.
    func sealFor(memberIds: [String], plaintext: String) async -> [(String, String)]? {
        guard available, !memberIds.isEmpty else { return nil }
        do {
            guard let deviceId = session.deviceId,
                  let me = await keys.privates(deviceId: deviceId)
            else { return nil }

            // Nobody but us in the recipient list means the caller could not work out who
            // the message is for — a group, where the client holds no full membership. A
            // send like that would seal to this account's own devices and post something
            // the rest of the room could never read, so it goes out in the clear instead,
            // which is what a group message already is.
            guard memberIds.contains(where: { $0 != me.userId }) else { return nil }

            var targets: [String] = []
            for userId in memberIds {
                targets.append(contentsOf: await devicesOf(userId).keys)
            }
            targets = targets.filter { $0 != deviceId }
            guard !targets.isEmpty else { return nil }

            // A claim spends a one-time prekey from every device it returns, so
            // it asks only about the ones with no session yet. A conversation
            // that has been running a while claims nothing at all.
            var strangers: [String] = []
            for target in targets {
                if await store.loadSession(target) == nil { strangers.append(target) }
            }
            var bundles: [String: KeyBundle] = [:]
            if !strangers.isEmpty {
                let claimed = try await repo.claimKeys(userIds: memberIds, deviceIds: strangers)
                bundles = Dictionary(
                    claimed.bundles.map { ($0.deviceId, $0) },
                    uniquingKeysWith: { first, _ in first }
                )
            }

            var envelopes: [(String, String)] = []
            for target in targets {
                let bundle = bundles[target]
                let envelope: String? = await store.withSession(target) { existing in
                    let start = existing ?? bundle.flatMap { try? Cipher.beginSession(bundle: $0) }
                    guard let start else { return (nil, nil) }
                    let sealed = Cipher.sealWith(start, plaintext, sender: me)
                    return (sealed?.session, sealed?.envelope)
                }
                if let envelope { envelopes.append((target, envelope)) }
            }

            return envelopes.isEmpty ? nil : envelopes
        } catch {
            return nil
        }
    }

    // ── opening ──────────────────────────────────────────────────────────────

    /// What this device can read of an encrypted message.
    ///
    /// Asked of the local store first, and that is not an optimisation: a
    /// ratchet destroys a message key as it uses it, so a ciphertext opens
    /// exactly once on this device, ever. What was written down the first time
    /// is the only copy that survives a relaunch.
    ///
    /// `authorId` is the server's word for who wrote it, and the signature has
    /// to agree — a sealed body lifted from one message and hung under another
    /// name fails here rather than being shown under the wrong face.
    func open(_ messageId: String, _ ciphertext: String?, authorId: String?) async -> String? {
        if let known = await store.recall(messageId) { return known }
        guard let ciphertext, let authorId,
              let claim = Cipher.sealedSender(ciphertext),
              claim.userId == authorId,
              let deviceId = session.deviceId,
              let me = await keys.privates(deviceId: deviceId),
              let senderKey = await identityKey(of: claim.userId, device: claim.deviceId)
        else { return nil }

        let read: Cipher.Read? = await store.withSession(claim.deviceId) { existing in
            let result = Cipher.openSealed(
                ciphertext,
                session: existing,
                me: me,
                senderIdentityKey: senderKey,
                expectedAuthorId: authorId
            )
            return (result?.session, result)
        }
        guard let read else { return nil }

        // Written down before anything else. A message that displays once and is
        // blank after a relaunch is worse than one that never showed.
        await store.remember(messageId, read.plaintext)

        // And only then is the prekey that started this session spent. In the
        // other order, a crash between the two would leave a message nobody can
        // ever read.
        if let consumed = read.consumedPreKeyId { await keys.consumePreKey(consumed) }
        return read.plaintext
    }

    /// A sender knows what it said; it should not have to open its own copy to prove it.
    func rememberOwn(_ messageId: String, _ plaintext: String) async {
        await store.remember(messageId, plaintext)
    }

    /// A deleted message leaves nothing readable behind on this device either.
    func forget(_ messageId: String) async {
        await store.forget(messageId)
    }
}

File: apps/webapp/src/lib/e2e.ts
import { ENCRYPTED_NOTICE, MESSAGE_FORMATS, chooseFormat } from '@yappy/shared';
import { api, currentDeviceId } from './api';
import type { RecipientBundle } from './cipher';
import { beginSession, openSealed, readFormats, sealWith, sealedSender } from './cipher';
import { e2eAvailable } from './e2eFlags';
import { consumePreKey, loadIdentity } from './keys';
import { recall, remember } from './plaintext';
import { loadSession, withSession } from './sessions';

/**
 * The session layer: who a message gets sealed to, what this device can open,
 * and where the answer is kept.
 *
 * The cipher is in `cipher.ts` and the ratchet under it in `ratchet.ts`. What
 * lives here is everything that has to touch the network or the disk: claiming
 * the prekeys that start a session, fetching the identity key a signature is
 * checked against, holding one lock per device so two sends cannot step the
 * same ratchet twice, and writing down what a message said — because with a
 * ratchet, that is the only copy that survives.
 */

/*
 * The flags themselves live in `e2eFlags.ts`, which imports nothing. Anything
 * that only needs to *ask* whether a room is private should import them from
 * there — reaching them through this module pulls the whole cipher with it.
 * Re-exported here so the existing callers read the same way.
 */
export { e2eAvailable, isPrivate, setPrivate } from './e2eFlags';

export interface Envelope {
  deviceId: string;
  ciphertext: string;
}

// ─── the identity keys of everybody else ─────────────────────────────────────

/** One device, as the directory describes it and this client believes it. */
interface DeviceProfile {
  identityKey: string;
  /** Verified against that key, or the oldest format if it did not verify. */
  formats: number[];
}

interface DirectoryEntry {
  fetchedAt: number;
  byDevice: Map<string, DeviceProfile>;
}

const directory = new Map<string, DirectoryEntry>();
/** Long enough that reading a screenful of history is one request per person. */
const DIRECTORY_TTL = 5 * 60 * 1000;

/**
 * Everything the directory says about one person's devices.
 *
 * Refetched when a message names a device the cache has never seen — that is
 * exactly what somebody adding a phone looks like from here, and the
 * alternative is that their first message from it is permanently unreadable.
 */
async function devicesOf(userId: string, wanted?: string): Promise<Map<string, DeviceProfile>> {
  const hit = directory.get(userId);
  if (hit && (wanted === undefined || hit.byDevice.has(wanted))) {
    if (Date.now() - hit.fetchedAt < DIRECTORY_TTL) return hit.byDevice;
  }
  try {
    const res = await api<{
      devices: Array<{
        deviceId: string;
        identityKey: string;
        formats: string | null;
        formatsSignature: string | null;
      }>;
    }>(`/keys/user/${userId}`);
    const byDevice = new Map(
      res.devices.map((d) => [
        d.deviceId,
        {
          identityKey: d.identityKey,
          // Checked here, once, rather than at each use: what the rest of
          // this file sees is a list the device itself signed.
          formats: readFormats(d.formats, d.formatsSignature, d.identityKey),
        },
      ]),
    );
    directory.set(userId, { fetchedAt: Date.now(), byDevice });
    return byDevice;
  } catch {
    return hit?.byDevice ?? new Map();
  }
}

/** The identity key a device publishes, which is what its signatures are checked against. */
async function identityKeyOf(userId: string, deviceId: string): Promise<string | null> {
  return (await devicesOf(userId, deviceId)).get(deviceId)?.identityKey ?? null;
}

// ─── sealing ─────────────────────────────────────────────────────────────────

/**
 * Bundles for the devices this one has never spoken to.
 *
 * Every claim consumes a one-time prekey from every device it returns, so it
 * asks only about the devices that still need one. A conversation that has been
 * running for a while claims nothing at all: the ratchet has taken over, and
 * the pool is left for the people who have not started yet.
 */
async function bundlesFor(userIds: string[], deviceIds: string[]): Promise<RecipientBundle[]> {
  if (deviceIds.length === 0) return [];
  const res = await api<{ bundles: RecipientBundle[] }>('/keys/claim', {
    method: 'POST',
    body: { userIds, deviceIds },
  });
  return res.bundles;
}

/**
 * What a private send puts on the wire.
 *
 * `content` carries the notice rather than the message — see
 * `messages.isEncrypted` on the server for why it is not left empty — and the
 * body goes one copy per device, each under its own ratchet.
 *
 * Own devices included — a message sent from a laptop has to be readable on the
 * phone — but not the device doing the sending, which cannot hold a ratchet
 * session with itself and writes down what it said instead.
 *
 * Returns null when there is nobody to encrypt to, which is not an error: it
 * means every recipient device is one this build has no usable keys for, and
 * the caller should send in the clear rather than post something nobody can
 * read.
 */
export async function sealFor(
  memberIds: string[],
  plaintext: string,
): Promise<{ content: string; envelopes: Envelope[]; plaintext: string } | null> {
  if (!e2eAvailable()) return null;
  try {
    const identity = await loadIdentity();
    if (!identity || identity.deviceId !== currentDeviceId()) return null;

    /**
     * Nobody but us in the recipient list means the caller could not work out who
     * the message is for — a group, where the client holds no full membership. A
     * send like that would seal to this account's own devices and post something
     * the rest of the room could never read, so it goes out in the clear instead,
     * which is what a group message already is.
     */
    if (!memberIds.some((id) => id !== identity.userId)) return null;

    const sender = {
      deviceId: identity.deviceId,
      userId: identity.userId,
      identityPrivate: identity.identityPrivate,
    };

    /**
     * Who exists, and what each of them can read.
     *
     * A device that speaks nothing this build can write gets no copy at all.
     * That is better than the alternative — a message sealed in a format the
     * recipient cannot parse is indistinguishable from a corrupted one, and
     * the ratchet has already thrown away the key by the time anybody notices.
     */
    const devices: Array<{ deviceId: string; format: number }> = [];
    for (const userId of memberIds) {
      for (const [deviceId, profile] of await devicesOf(userId)) {
        const format = chooseFormat(MESSAGE_FORMATS, profile.formats);
        if (format !== null) devices.push({ deviceId, format });
      }
    }
    /**
     * Everybody's devices except this one.
     *
     * A ratchet cannot talk to itself: one session record holds one sending
     * chain and one receiving chain, and a device sealing to its own session
     * would step the first and then try to read the message with the second.
     * The sender's own *other* devices are genuine strangers and get their
     * copies; the sending device writes down what it said instead, which the
     * ratchet already forces it to do for everything else.
     */
    const targets = devices.filter((d) => d.deviceId !== sender.deviceId);
    const unknown: string[] = [];
    for (const { deviceId } of targets) {
      if (!(await loadSession(deviceId))) unknown.push(deviceId);
    }
    const bundles = new Map(
      (await bundlesFor(memberIds, unknown)).map((b) => [b.deviceId, b] as const),
    );

    const envelopes: Envelope[] = [];
    for (const { deviceId, format } of targets) {
      const envelope = await withSession(deviceId, async (existing) => {
        let session = existing;
        if (!session) {
          const bundle = bundles.get(deviceId);
          if (!bundle) return { session: null, result: null };
          try {
            session = beginSession(bundle);
          } catch {
            // A device whose prekey does not verify. Dropped on its own — that
            // is one bad device, and everybody else is still owed their copy.
            return { session: null, result: null };
          }
        }
        const sealed = sealWith(session, plaintext, sender, format);
        return { session: sealed?.session ?? null, result: sealed?.envelope ?? null };
      });
      if (envelope) envelopes.push({ deviceId, ciphertext: envelope });
    }

    return envelopes.length > 0 ? { content: ENCRYPTED_NOTICE, envelopes, plaintext } : null;
  } catch {
    return null;
  }
}

// ─── opening ─────────────────────────────────────────────────────────────────

/**
 * What this device can read of an encrypted message.
 *
 * Asked of the local store first, and that is not an optimisation. A ratchet
 * destroys a message key as it uses it, so a ciphertext opens exactly once on
 * this device, ever. What was written down the first time is the only copy that
 * survives a reload — see `plaintext.ts`.
 *
 * `authorId` is the server's word for who wrote it, and the signature has to
 * agree with it: a sealed body lifted from one message and hung under another
 * name fails here rather than being shown under the wrong face.
 */
export async function decrypt(
  messageId: string,
  ciphertext: string | null | undefined,
  authorId: string,
): Promise<string | null> {
  const known = await recall(messageId);
  if (known !== null) return known;
  if (!ciphertext) return null;

  const claim = sealedSender(ciphertext);
  if (!claim || claim.userId !== authorId) return null;

  const identity = await loadIdentity();
  if (!identity || identity.deviceId !== currentDeviceId()) return null;

  const senderKey = await identityKeyOf(claim.userId, claim.deviceId);
  if (!senderKey) return null;

  const me = {
    deviceId: identity.deviceId,
    signedPreKeyId: identity.signedPreKeyId,
    signedPreKeyPrivate: identity.signedPreKeyPrivate,
    preKeys: identity.preKeys,
  };

  const opened = await withSession(claim.deviceId, async (session) => {
    const result = openSealed(ciphertext, session, me, senderKey, authorId);
    return { session: result?.session ?? null, result };
  });
  if (!opened) return null;

  // Written down before anything else. A message that displays once and is
  // blank after a reload is worse than one that never displayed.
  await remember(messageId, opened.plaintext);

  // And only then is the prekey that started this session spent. In the other
  // order, a crash between the two would leave a message nobody can ever read.
  if (opened.consumedPreKeyId !== null) await consumePreKey(opened.consumedPreKeyId);
  return opened.plaintext;
}

/** A sender knows what it said; it should not have to open its own copy to prove it. */
export async function rememberOwn(messageId: string, plaintext: string): Promise<void> {
  await remember(messageId, plaintext);
}
